torch/csrc/jit/codegen/cuda/expr_simplifier.h

```diff
@@ -0,0 +1,536 @@
+// described as below:
+//
+// Theorem 3.1 (Division Theorem): for any integers a and b (b != 0), there
+// exist unique integers q and r such that:
+// 1) 0 <= r < |b|
+// 2) a = bq + r
+//
+// if we define a/b := q, and a%b := r, then we can write 2) as:
+//   a = (a/b)b + a%b
+// I will call the above equation "the fundamental division-with-remainder
+// equation" in later contexts.
+//
+// In this section and later sections, I will use \ for real, and / for integer
+// div. For example, 5\2 = 2.5, 5/2 = 2.
+//
+// Note that the a/b and a%b in number theory is not the same as those in C++,
+// as explained in later section "Implementations of Div and Mod".
+//
+// A very important concept in number theory is "congruence", as described
+// below:
+//
+// Definition 3.1: For c != 0, if (a-b)\c is an integer, then we say a is
+// congruent to b modulo c, written as a = b (mod c).
+//
+// Note that the "(mod k)" in the statement "a = b (mod k)" should be understood
+// as a qualifier for the statement "a = b" that modifies the meaning of the
+// original statement [P] to "[P] under the modular arithmetic". It should NOT
+// be understood as an operator operating on "b", i.e. "a = (b (mod k))".
+//
+// Theorem 3.2: The congruence is an equivalence relationship, that is:
+// - a = a (mod c)
+// - if a = b (mod c), then b = a (mod c)
+// - if a = b (mod c) and b = d (mod c), then a = d (mod c)
+//
+// Theorem 3.3: The congruence also has the following properties:
+// - a = b (mod c) iff a = b (mod (-c))
+// - if a = a' (mod c), b = b' (mod c), then
+//   - a + b = a' + b' (mod c)
+//   - ab = a'b' (mod c)
+// - if bd = bd' (mod c) and gcd(b, c) = 1, then d = d' (mod c)
+//
+// 4. Some More Theorems
+// =====================
+//
+// In this section, I will prove a few theorems that I didn't find in textbooks,
+// but still feel useful for us. The / and % are under the definition of number
+// theory, not under the definition of C++. The properties of / and % with the
+// definition of C++ will be revisited in later section "Properties of Div and
+// Mod Under Trunc Div".
+//
+// Bad news comes first:
+//
+// Theorem 4.1: Integer div is not associative:
+// 1) a*(b/c) != (a*b)/c
+// 2) a/(b*c) != (a/b)/c
+// 3) a/(b/c) != (a/b)*c != (a*c)/b
+// Proof: a*(b/c) != (a*b)/c because of the counter example 2*(3/2) != (2*3)/2
+// a/(b*c) != (a/b)/c because of the counter example 4/((-1)*5) != (4/(-1))/5
+// Note that 4/(-5) = 0, (-4)/5 = -1
+// a/(b/c) != (a/b)*c != (a*c)/b because of the counter example that
+// 5/(3/2), (5/3)*2, and (5*2)/3 are three different numbers
+//
+// Theorem 4.2: Integer div is NOT right distributive
+// Proof: Counter example: (1+1)/2 != 1/2 + 1/2
+//
+// Then comes good properties:
+//
+// Theorem 4.3: a%b = a'%b is equivalent to a = a' (mod b)
+// Proof: Direction ==>:
+// a%b = a'%b is equivalent to a-(a/b)b = a'-(a'/b)b
+// which is equivalent to (a-a')\b = (a/b-a'/b) = integer
```

Comment:
It might be simpler to cite the C99 standard here:

```
When integers are divided, the result of the / operator is the algebraic quotient with any fractional part discarded.

If the quotient a/b is representable, the expression (a/b)*b + a%b shall equal a.
```

So `a%b == a - (a/b)*b` and the statement is immediate.

Reply:
I think it should be `\b` here? Because `/b` always returns an integer, so `.../b is integer` is an always-true statement. And in this section 4, I am talking about the Euclidean div, not the truncation div, so C99 standard is not related here. But I added the quote to the C99 standard at section "6. Properties of Div and Mod Under Trunc Div"